tools/testutils.py

```diff
@@ -17,133 +17,160 @@
+    expected = "".join(expected.split()).upper()
     if len(received) < len(expected):
-        report_err(outputs["stderr"], "Received packet too short",
-                   len(received), "vs", len(expected))
+        log.error(f"Received packet too short {len(received)} vs {len(expected)}")
         return FAILURE
-    for i in range(0, len(expected)):
-        if expected[i] == "*":
+    for idx, val in enumerate(expected):
+        if val == "*":
             continue
-        if expected[i] != received[i]:
-            report_err(outputs["stderr"], "Received packet ", received)
-            report_err(outputs["stderr"], "Packet different at position", i,
-                       ": expected", expected[i], ", received", received[i])
-            report_err(outputs["stderr"], "Expected packet ", expected)
+        if val != received[idx]:
+            log.error(f"Received packet\n {received} ")
+            log.error(
+                f"Packet different at position {idx}: expected\n{val}\n\nreceived\n{received[idx]}"
+            )
+            log.error(f"Expected packet {expected}")
             return FAILURE
     return SUCCESS
 
 
-def open_process(verbose, args, outputs):
-    """ Run the given arguments as a subprocess. Time out after TIMEOUT
-        seconds and report failures or stdout. """
-    report_output(outputs["stdout"], verbose, "Writing", args)
+def is_tcp_port_in_use(port: int) -> bool:
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        return s.connect_ex(("localhost", port)) == 0
+
+
+def pick_tcp_port(default_port: int) -> int:
+    # Pick an available port.
+    port_used = True
+    while port_used:
+        default_port = random.randrange(1024, 65535)
+        port_used = is_tcp_port_in_use(default_port)
+
+    return default_port
+
+
+def open_process(args):
+    """Run the given arguments as a subprocess. Time out after TIMEOUT
+    seconds and report failures or stdout."""
+    log.info(f"Writing {args}")
     proc = None
 
     try:
-        proc = Popen(args,
-                     stdout=subprocess.PIPE,
-                     shell=True,
-                     stdin=subprocess.PIPE,
-                     stderr=subprocess.PIPE,
-                     universal_newlines=True)
+        proc = subprocess.Popen(
+            args,
+            stdout=subprocess.PIPE,
+            shell=True,
+            stdin=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+            preexec_fn=os.setsid,
+        )
     except OSError as e:
-        report_err(outputs["stderr"], "Failed executing: ", e)
+        log.error(f"Failed executing:\n{e}")
     if proc is None:
         # Never even started
-        report_err(outputs["stderr"], "Process failed to start")
+        log.error("Process failed to start")
     return proc
 
 
-def run_process(verbose, proc, timeout, outputs, errmsg):
-
+def run_process(proc, timeout, errmsg):
```

Comment:
The name is a bit misleading, it suggest the process is executed, but in fact only its outputs are consumed.